nupic/research/frameworks/pytorch/imagenet/mixins/lr_range_test.py

```diff
@@ -0,0 +1,111 @@
+
+import os
+from copy import deepcopy
+
+from pandas import DataFrame
+from ray.tune.utils import flatten_dict
+
+from nupic.research.frameworks.pytorch.imagenet import mixins
+from nupic.research.frameworks.pytorch.lr_scheduler import LinearLRScheduler
+
+
+class LRRangeTest(mixins.LogEveryLoss):
+    """
+    Mixin for the LR-range test defined in section 4.1 of "A Disciplined Approach to
+    Neural Network Hyper-Parameters"
+        - https://arxiv.org/pdf/1803.09820.pdf
+
+    Herein, a min_lr and max_lr are set, and training proceeds for a small number of
+    epochs (1-3) while the learning rate is linearly increased. Generally, the point
+    at which the training loss begins to curve upwards and increases while the
+    validation/test loss still decreases, is considered to be a reasonable choice
+    for your max_lr in a cyclical lr-schedule. The same author recommends using 10-20
+    times lower this amount for your min_lr.
+    """
+
+    # Use the following name when `save_lr_test_to_csv=True`.
+    csv_save_name = "lr_range_test.csv"
+
+    def setup_experiment(self, config):
+        """
+        :param config:
+            - epochs: number of epochs in training (recommended 1-3)
+            - epochs_to_validate: will be overridden to include all epochs
+            - save_lr_test_to_csv: whether to save results to csv; must be accompanied
+                                   by 'logdir'
+            - logdir: directory to save test results
+            - lr_scheduler_class: automatically overridden to LinearLRScheduler
+            - lr_scheduler_args: args for the linear-schedule
+                - min_lr: starting learning rate
+                - max_lr: ending learning rate
+        """
+        # Ensure all epochs get validated.
+        assert "epochs" in config
+        config["epochs_to_validate"] = range(-1, config["epochs"])
+
+        # Init saving of test results.
+        self.save_lr_test_to_csv = config.get("save_lr_test_to_csv", False)
+        if self.save_lr_test_to_csv:
+            assert "logdir" in config, (
+                "The logdir must be specified to save the lt-range test resutls.")
+            self.logdir = config["logdir"]
+
+        # Save config for later - used to aggregate results.
+        self._config = deepcopy(config)
+        super().setup_experiment(config)
+
+    @classmethod
+    def create_lr_scheduler(cls, config, optimizer, total_batches):
+
+        assert "lr_scheduler_args" in config
+
+        lr_scheduler_class = LinearLRScheduler
+        lr_scheduler_args = config["lr_scheduler_args"]
+
+        lr_scheduler_args.update(epochs=config["epochs"])
+        lr_scheduler_args.update(steps_per_epoch=total_batches)
+
+        return lr_scheduler_class(optimizer, **lr_scheduler_args)
+
+    def post_batch(self, *args, **kwargs):
+        """Increase lr after every batch."""
+        self.lr_scheduler.step()
+        super().post_batch(*args, **kwargs)
```

Comment:
Similar question: super() should probably be at the beginning?

Reply:
Fair point. I've fixed both places where this occurs.